docs/smtchecker.rst

```diff
@@ -0,0 +1,737 @@
+.. _formal_verification:
+
+##################################
+SMTChecker and Formal Verification
+##################################
+
+Using formal verification it is possible to perform an automated mathematical
+proof that your source code fulfills a certain formal specification.
+The specification is still formal (just as the source code), but usually much
+simpler.
+
+Note that formal verification itself can only help you understand the
+difference between what you did (the specification) and how you did it
+(the actual implementation). You still need to check whether the specification
+is what you wanted and that you did not miss any unintended effects of it.
+
+Solidity implements a formal verification approach based on
+`SMT <https://en.wikipedia.org/wiki/Satisfiability_modulo_theories>`_ and
+`Horn <https://en.wikipedia.org/wiki/Horn-satisfiability>`_ solving.
+The SMTChecker module automatically tries to prove that the code satisfies the
+specification given by ``require`` and ``assert`` statements. That is, it considers
+``require`` statements as assumptions and tries to prove that the conditions
+inside ``assert`` statements are always true.  If an assertion failure is
+found, a counterexample may be given to the user showing how the assertion can
+be violated. If no warning is given by the SMTChecker for a property,
+it means that the property is safe.
+
+The other verification targets that the SMTChecker checks at compile time are:
+
+- Arithmetic underflow and overflow.
+- Division by zero.
+- Trivial conditions and unreachable code.
+- Popping an empty array.
+- Insufficient funds for a transfer.
+
+The potential warnings that the SMTChecker reports are:
+
+- ``<failing  property> happens here.``. This means that the SMTChecker proved that a certain property fails. A counterexample may be given, however in complex situations it may also not show a counterexample. This result may also be a false positive in certain cases, when the SMT encoding adds abstractions for Solidity code that is either hard or impossible to express.
+- ``<failing property> might happen here``. This means that the solver could not prove either case within the given timeout. Since the result is unknown, the SMTChecker reports the potential failure for soundness. This may be solved by increasing the query timeout, but the problem might also simply be too hard for the engine to solve.
+
+It is currently an experimental feature, therefore in order to use it you need
+to enable it via :ref:`a pragma directive<smt_checker>`.
```

Comment:
Maybe this should be the first thing to mention, so maybe should be moved to top? But instead of `It` in that case you can use `Formal verification ...` or something similar. But, that is just my feeling, maybe it is fine as is...

Reply:
@mijovic I would prefer to leave it there because we first explain what it does, and then how to use it. I can still change it if you strongly feel like it should come before.

Reply:
Let's leave it like it is, I didn't have strong opinion on that, just felt like it could be moved. But, since it is not a huge paragraph before I think it is completely fine as is.